### ISSUE TITLE
translation not working with globalResources

### ISSUE THREAD

**Issue by tbnovaes** [aurelia/i18n#192]
**I'm submitting a bug report**

* **Library Version:**
aurelia-18n@1.2.3

**Please tell us about your environment:**
* **Operating System:**
OSX 10.12.2 | Windows 10

* **Node Version:**
6.2.2

* **NPM Version:**
3.10.6

* ** Aurelia-CLI **
0.23.0

* **Browser:**
all

* **Language:**
ESNext

**Current behavior:**
if I have a HTML file translated using i18n it works just fine. If I add this file to `globalResources`, and use `.feature()` to load it, the translation does not work but the translation keys are displayed instead.

**Expected/desired behavior:**
That translations work with the globalResources.

* **What is the motivation / use case for changing the behavior?**
Being able to globalize custom components that are used in many places around the app, without losing the translation.

**Comment by zewa666** [comment]
Hi @tbnovaes. Interesting never thought of this to be honest. Could you create a minimal example to take a look at it?

**Comment by tbnovaes** [comment]
@zewa666 I'll try to reproduce it in a new project, and let you know the repo.

**Comment by peterver** [comment]
+1 

I bumped into the same issue today, i18n is initialized correctly though, it's the value converters that don't appear to work for me.

if i try the following : 

```
<input type="text" placeholder="${'hello.world' | t}" />
```

it doesnt work. But the following does work : 

```
<input type="text" t="[placeholder]hello.world" />
```

**Comment by peterver** [comment]
After turning on debug of aurelia-i18n i found something rather peculiar : 

![screen shot 2017-01-23 at 03 55 22](https://cloud.githubusercontent.com/assets/821415/22189896/e3efc21c-e11f-11e6-90e4-5a972a01b7c7.png)

Apparently calls for translations are being done before i18n has initialized? Since value converters are one-time binds that would explain why the full-fledged two-way bound 't' does work

**Comment by peterver** [comment]
This is most likely not the best fix possible, but a workaround for now would be : 

```
.plugin('aurelia-i18n', (instance) => {
    ...
    **instance.i18next.on('loaded', function () {
        aurelia.setRoot('app');
    });**
    ...
});
```

reference in i18next spec : [on-loaded](http://i18next.com/docs/api/#on-loaded)

**Comment by zewa666** [comment]
@peterver thanks for further investigation, @tbnovaes hope you had a chance to build a small example so we can take a look at your specific case

**Comment by peterver** [comment]
@zewa666 you're welcome, feel free to mention me in any future comments as i am most likely dealing with the same issue as @tbnovaes in my stack

**Comment by tbnovaes** [comment]
@zewa666 yes, I did build a new project, but I'm trying to add the same issue on it, because the behavior I'm having in the sample project is different from what I'm having in my real application.

**Comment by zewa666** [comment]
@tbnovaes awesome. Maybe @peterver you can help out with that as well

**Comment by peterver** [comment]
what exactly is it you require @zewa666 ? :]

**Comment by zewa666** [comment]
@peterver What would really help me to debug issues is to have a small example at hand. Most of the time the issue is a combination of certain steps. How is the data defined, what were the init options, and then the specific use case itself. So instead of me trying to second guess what happened, a dummied down example greatly reduces the time and effort I have, and thus the bugfix can come in quicker :)

You can either do it with a gist.run (https://gist.run/?id=ffe167be3a1d8d5a0526a415febe85a1) or simply create a repo and share the link with a minimal CLI example. Whichever ways you guys prefer

**Comment by peterver** [comment]
@zewa666 Sorry for the late reply, i'll put in some work tonight on a dumbed down version of my stack ( koa serverside and aurelia clientside ) to act as an example :].

**Comment by peterver** [comment]
@zewa666 You can take a look at the following folder ( i didn't even have to use global resources ) https://github.com/peterver/aurelia-i18n-globalresource-issue

**Comment by zewa666** [comment]
Well I've tried everything to bundle your app locally on windows, but seems there are several issues with the package.json, not all dev-deps are included and the develop:app script fails with a missing preset env.

Nevertheless I think the issue could be in this line: https://github.com/peterver/aurelia-i18n-globalresource-issue/blob/master/src/public/app/main.js#L19

the setup method returns a promise which should be returned in order to guarantee proper timings. See the docs http://aurelia.io/hub.html#/doc/article/aurelia/i18n/latest/i18n-with-aurelia/3.

So please try out:

```
return instance.setup({...});
```

**Comment by peterver** [comment]
@zewa666 ok well now i feel stupid :'), adding the return statement definitely works for me, even the global resources work perfectly

**Comment by zewa666** [comment]
nothing to feel stupid about, happened to myself more then often, which is a reason why I now setup projects always by coping the section from the hub 😆 

@tbnovaes can you confirm that you're having the same issue?

**Comment by tbnovaes** [comment]
@peterver @zewa666 no in my case I'm returning a promise. But when I try to reproduce the error in a new project, I can not. It is workig just fine. So I'll keep investigating, if I'm able to put it in a new project than I can reopen the case. Otherwise probably is something that I'm doing wrong some place else.

**Comment by zewa666** [comment]
@tbnovaes allright, I'm going to close this issue then and let us reopen it if you find out something new

**Comment by timfish** [comment]
I'm seeing this too using `i18next-xhr-backend`. 

In all my custom elements added through `globalResources`, the `i18n` attribute doesn't work, whereas the `t` binding behaviour works fine.

```typescript
  aurelia.use
    .standardConfiguration()
    .plugin('aurelia-i18n', async (instance: I18N) => {
      instance.i18next.use(Backend);
      await instance.setup({
        attributes: ['t', 'i18n'],
        backend: {
          loadPath: './locales/{{lng}}/{{ns}}.json',
        },
        debug: environment.hasDevSwitch,
        fallbackLng: 'en',
        lng: locale.settings.localeName,
        interpolation: {
          defaultVariables: {
            appName: environment.appName
          }
        }
      });

      const router: Router = aurelia.container.get(Router);
      router.transformTitle = title => instance.tr(title);
    })
    // globalise resources from the resources directory
    .feature('resources');
```

**Comment by zewa666** [comment]
I dunno how await behaves here but have you tried returning the promise as mentioned above in the discussion?

**Comment by timfish** [comment]
Yes, even the following doesn't appear to work:

```typescript
  aurelia.use
    .standardConfiguration()
    .plugin('aurelia-i18n', async (instance: I18N) => {
      instance.i18next.use(Backend);

      const router: Router = aurelia.container.get(Router);
      router.transformTitle = title => instance.tr(title);

      return instance.setup({
        attributes: ['t', 'i18n'],
        backend: {
          loadPath: './locales/{{lng}}/{{ns}}.json',
        },
        debug: environment.hasDevSwitch,
        fallbackLng: 'en',
        lng: locale.settings.localeName,
        interpolation: {
          defaultVariables: {
            appName: environment.appName
          }
        }
      });
    })
    // globalise resources from the resources directory
    .feature('resources');
```

**Comment by zewa666** [comment]
Can you create a new issue with an example gist? It does somehow sound like a new thing

**Comment by timfish** [comment]
I'll create a repo based off a skeleton. My setup has been doing this for over 6 months. I've got loads of nasty `${'buttons.save' & t}` in my components back when I started.

**Comment by timfish** [comment]
[This repository](https://github.com/timfish/aurelia-i18n-issue192) reproduces the issue.

`some-component.html`
```html
<template>
    <span i18n="some-text"></span>
    <span>${'some-text' & t}</span>
</template>
```
Only displays the translation from the binding behaviour:

![image](https://cloud.githubusercontent.com/assets/1150298/23758159/1b699da2-04e9-11e7-9552-b6c69334f250.png)


I probably should have added it in two commits so you could see the changes I made from the typescript skeleton. Here's a summary:

- added `/resources/*`
- added `.feature('resources')` to `main.ts`
- added `<some-component>` to `welcome.html`
- `jspm install` `aurelia-i18n` and `i18next-xhr-backend`
- added `i18n` to `main.ts`
- added `./locales/en/translation.json`
- changed `"module": "system"` in `tsconfig.json`

**Comment by zewa666** [comment]
ok this is definitely a new bug. The issue is the aliasing of the TCustomAttribute. Changing `<span i18n="some-text"></span>` to `<span t="some-text"></span>` will fix it meanwhile. I'll have a look whats happening here

**Comment by AStoker** [comment]
Same issue here. Looking into it as well. For the sake of testing, I have a bare bones repo here: https://github.com/AStoker/i18nAttributeComparison

**Comment by AStoker** [comment]
So just to add some details here, I've been digging through the code, and this is what _appears_ to be the order of operations.  

1. Project loads up, and i18next gets "configured". In reality, it's not completely configured until Aurelia runs through it's `postTasks` queue (https://github.com/aurelia/i18n/blob/master/src/aurelia-i18n.js#L26). So the code that is responsible for registering the `i18n` alias must wait for the `postTask` queue to hit it. However, since the i18n plugin is loading up resources lazily, it tells Aurelia to continue on it's way, and Aurelia reports that the i18n plugin is ready (despite a promise waiting in the background).   
2. In the mean time, Aurelia will continue going through loading up its resources. This _includes_ loading up the `aurelia-i18n/t` resource (this is why the `t` attribute works, since it's loaded up directly as a resource, while `i18n` is an _alias_ for `t`, and isn't configured until later).  
3. Eventually, we get to the part where we load up our custom element (since it's defined as a global resource, it's loading up at the start). This (like how we loaded up the `aurelia-i18n/t` resource) is loaded up and compiled. At this point, we are still loading up resources, so Aurelia hasn't called our `postTasks` queue yet, meaning `i18n` isn't a registered alias. However, since we're now compiling our view, and we only have the `t` attribute registered, only the `t` attributes in our view get set up correctly.  
4. Our resources are all done loaded, and Aurelia finally calls the `postTasks` queue to be flushed. At this point, we finally register the `i18n` attribute as an alias for the `t` attribute (far after we need it to be).  

Not sure of a solution yet, but hopefully this sheds some light on what the issue _possibly_ is.

**Comment by zewa666** [comment]
That's as far as I was, so it's good to see we're on the same track. Utilizing preTask didn't help either, so I have to fiddle around and see how to let the aliasing part be executed earlier

**Comment by AStoker** [comment]
Essentially, we would have to somehow piggyback off of when the `t` attribute is created, since Aurelia is going through all the resources at once. We can't wait for it to finish processing the resources, since it's too late for globals. Not sure exactly how to accomplish that, or if it's even possible/a good idea.

**Comment by tbnovaes** [comment]
@zewa666 that's why I could not reproduce the issue on the example project I was creating. In my real code I was using `i18n` alias, but in the testing project I was using `t`, changing the `i18n` to `t` in my code made it work as expected.

**Comment by zewa666** [comment]
@tbnovaes, @timfish sry for not coming back to you guys in a while, but meantime, with the help of @AStoker, we've found the reason for the issue.

In order to properly fix this, we need to add a new feature to aurelias templating repo which is introduced with this PR https://github.com/aurelia/templating/pull/535.
Once this is done I'll update the docs and put some instructions on how to leverage attribute aliases properly for the i18n plugin

**Comment by timfish** [comment]
Great to have this fixed. Once we have a new `aurelia-i18n` release, will there be any changes required or will the aliases just start working again in global resources?

**Comment by AStoker** [comment]
You'll have to configure things slightly @timfish (It's probably worth documenting how to set up aliases, @zewa666)
But what you'll have to do is this:
``` javascript
// Make sure you import the TCustomAttribute
import {TCustomAttribute} from 'aurelia-i18n';
...
export function configure(aurelia) {
    // Call the static helper method on the TCustomAttribute to configure aliases. Give it an array of aliases to use. Duplicates are fine (so in this case, `t` is redundant)
    TCustomAttribute.configureAliases(['t', 'i18n']);
    ...
}
```
Note, you can also use this for your `t-params` custom attribute too. Just pull in the `TParamsCustomAttribute` and do the exact same thing you did for `TCustomAttribute`.
So the `attributes` configuration option is no longer needed in your setup.

**Comment by timfish** [comment]
@AStoker Nice! Thanks.

I wasn't even aware `i18n` was an alias until stumbling across this issue. I'd decided to use it over `t` as it seems a little more explicit for what it's doing.

**Comment by AStoker** [comment]
I had the exact same situation :P

**Comment by tbnovaes** [comment]
Thank you! @AStoker @zewa666. Glad to here that this one is fixed!

**Comment by timfish** [comment]
Tried this today with the new releases and `TCustomAttribute.configureAliases(['t', 'i18n'])` in the top of configure (as per @AStoker example) appears to break things and stop the app from loading.

I updated all aurelia dependencies:
`jspm install aurelia-animator-css aurelia-binding aurelia-bootstrapper aurelia-computed aurelia-dialog aurelia-event-aggregator aurelia-fetch-client aurelia-framework aurelia-history-browser aurelia-i18n aurelia-loader-default aurelia-logging-console aurelia-pal-browser aurelia-polyfills aurelia-router aurelia-templating-binding aurelia-templating-resources aurelia-templating-router aurelia-testing`

On run, first error:
![image](https://cloud.githubusercontent.com/assets/1150298/24398557/2d332c40-13aa-11e7-825c-8588da1b908e.png)

But then it also fails to load `compose`, `if`, `with`, `repeat` etc from `aurelia-templating-resources@1.3.1` which stops the app from loading.

**Comment by AStoker** [comment]
@timfish, sorry, my mistake. I thought we could remove the attributes option, but we need to leave it in there for now. 
Our fix was an initial one, we'd like to find a more elegant solution, but this provided us with a way to get roadblocks out of the way for people. So go ahead and just add your `attributes: ['t', 'i18n'],` code back in, and see if that works.

**Comment by timfish** [comment]
Even with just ` TCustomAttribute.configureAliases(['t', 'i18n'])` added I get the above. It doesn't seem to matter if `attributes: ['t', 'i18n'],` is there or not.

I've cleaned out all the forks listed by `jspm` (although none were aurelia libs) and tried reinstalling them all from scratch but it didn't help, so I'm digging into the exception.

One of the expressions appears to be `undefined`:


![image](https://cloud.githubusercontent.com/assets/1150298/24401992/f56dec16-13b6-11e7-80e6-9095419213f4.png)

It appears to occur when loading:
`./jspm_packages/npm/aurelia-dialog@1.0.0-beta.3.0.1/ai-dialog-footer.js`

On HTML  node:
`<button type="button" class="btn btn-default" repeat.for="button of buttons" click.trigger="close(button)">\${button}</button>`

That happens because `instruction.attrName` is `'repeat'` and it doesn't exist on `instruction.attributes`:

![image](https://cloud.githubusercontent.com/assets/1150298/24402239/f17f51e8-13b7-11e7-98ee-bb0b65edff19.png)

Should I report this as a bug elsewhere?

**Comment by AStoker** [comment]
I'll be at work in about an hour where I can dig a little more into this. Make sure you have the latest version of the templating engine too (1.4.0)

**Comment by timfish** [comment]
Thanks.

Yes, the exception is coming from within `aurelia-templating@1.4.0`
```
TypeError: Cannot read property 'attrToRemove' of undefined
    at ViewCompiler._compileElement (file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:2839:27)
    at ViewCompiler._compileNode (file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:2548:23)
    at ViewCompiler._compileNode (file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:2570:33)
    at ViewCompiler.compile (file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:2517:12)
    at ViewCompiler._compileElement (file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:2722:28)
    at ViewCompiler._compileNode (file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:2548:23)
    at ViewCompiler._compileNode (file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:2570:33)
    at ViewCompiler.compile (file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:2517:12)
    at file:///C:/Users/.../jspm_packages/npm/aurelia-templating@1.4.0/aurelia-templating.js:3218:50
```

**Comment by AStoker** [comment]
Just for kicks and giggles, does it happen when you don't use the `configureAliases` function?

**Comment by timfish** [comment]
No, unfortunately not . It works fine when I comment out the call to `configureAliases `

**Comment by AStoker** [comment]
Okay, that at least rules out us hosing the templating engine 😛

**Comment by timfish** [comment]
> No, unfortunately not

Or fortunately not 😂

**Comment by timfish** [comment]
@AStoker Did you manage to reproduce this your end?

**Comment by AStoker** [comment]
@timfish, sorry I'm just getting back to you, a few other things came up hard and fast. I just tested these changes out on my local machine, and they seem to work fine. Are you still getting this error? Can you recreate the error in a basic skeleton project?

**Comment by timfish** [comment]
@AStoker thank. I'll get something together

**Comment by AStoker** [comment]
Let me correct my previous statement, I don't get any errors in the latest. I still have a few things missing, so it's possible that I'm not set up correctly yet...

**Comment by timfish** [comment]
It looks like it's caused by Aurelia dialog so that'll be the first thing I try.

**Comment by AStoker** [comment]
Yah, my little hiccup was just some cache stuff. It's working on my end.

**Comment by timfish** [comment]
The exceptions loading `aurelia-dialog` are because most of the built-in attributes are not in the `ViewResources` and it's one of the first modules that gets loaded. Remove `aurelia-dialog` and it dies when try and use the attributes later.

It took a while but I finally found the culprit: 

- Take the basic Typescript skeleton
- Add i18n and use `configureAliases`
- `jspm install reflect-metadata`
- Add `import 'reflect-metadata';` to the top of `main.ts`

Or this repo: https://github.com/timfish/aurelia-i18n-issue192

I still have no idea why is causes the issue but it's late here and I need sleep!
